### PR TITLE
Use hiera data

### PR DESCRIPTION
**Pull request by ekohl**

### CODE REVIEW

**Comment by wyardley** [comment]
I think it would definitely be useful if we had a working prototype of "data in modules" if that's the direction things are heading.

**Comment by bastelfreak** [comment]
data in modules \o/

**Comment by ekohl** [comment]
Using .yml extensions while hiera looks for .yaml makes rspec sad. Should be better now.

File: data/ArchLinux.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+nginx::daemon_user: http
```
Comment (bastelfreak) [inline data/ArchLinux.yaml]: Archlinux \o/

I'm pretty sure filename should be Archlinux.yaml

Comment (ekohl) [reply data/ArchLinux.yaml]: I copied it from ```params.pp``` where it's also ```ArchLinux```. Might be good to verify there's tests for it.

**Comment by ekohl** [comment]
This is still WIP because we'll need to bump the minimal puppet version to something 4.9+ or 4.10+. puppetlabs-ntp has 4.9.4 as minimal version and uses hiera 5.

File: manifests/params.pp [hunk 2]
```diff
@@ -1,124 +0,0 @@
-# Class: nginx::params
-# ====================
-#
-# nginx default settings and according to operating system
-#
-class nginx::params {
-  ### Operating System Configuration
-  ## This is my hacky... no hiera system. Oh well. :)
```
Comment (ekohl) [inline manifests/params.pp]: Fixed! :)

File: data/Solaris.yaml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+---
+nginx::daemon_user: webservd
+nginx::package_name: undef
```
Comment (ekohl) [inline data/Solaris.yaml]: I'm pretty sure this is not the way to unset it and need a better solution.

File: metadata.json [hunk 4]
```diff
@@ -20,7 +20,7 @@
   "requirements": [
     {
       "name": "puppet",
-      "version_requirement": ">= 4.7.0 < 6.0.0"
+      "version_requirement": ">= 4.9.4 < 6.0.0"
```
Comment (bastelfreak) [inline metadata.json]: We had a discussion yesterday on the IRC about lowest version of puppet in modules with data-in-modules. We decided to go with 4.10.9 as there were soo many bug fixes for the hiera implementation.

**Comment by ekohl** [comment]
Updated

**Comment by bastelfreak** [comment]
Should we merge this now as is?

**Comment by ekohl** [comment]
Should we also consider this a major version bump? I'd like to get some more reviews on this.

**Comment by bastelfreak** [comment]
yeah a major bump is probably needed.

**Comment by dhollinger** [comment]
puppet-rsyslog uses data in modules if you want some other examples

**Comment by ekohl** [comment]
I based it on theforeman/tftp wich also uses data in modules. I'm fairly confident it works but maybe the bigger point is that I want an ACK that this is something we want in our modules and maybe a new goal to work towards.

**Comment by bastelfreak** [comment]
@ekohl can you rebase this? I would like to the the 1.0.0 release next and include this. Bumping lowest puppet version boundary to 4.10.10 is fine.

**Comment by ekohl** [comment]
Rebased

File: manifests/init.pp [hunk 5]
```diff
@@ -26,33 +26,33 @@
 # }
 class nginx (
   ### START Nginx Configuration ###
-  $client_body_temp_path                                     = $::nginx::params::client_body_temp_path,
-  Boolean $confd_only                                        = false,
-  Boolean $confd_purge                                       = false,
-  $conf_dir                                                  = $::nginx::params::conf_dir,
+  Stdlib::Absolutepath $client_body_temp_path                = undef,
```
Comment (alexjfisher) [inline manifests/init.pp]: `undef` isn't a valid default for the data type (and the default is overridden in the module data anyway).

**Comment by ekohl** [comment]
Moved a few more parameters and added a few more data types.

**Comment by wyardley** [comment]
I'm going to restart just in case, but looks like multiple tests are failing?

**Comment by bastelfreak** [comment]
@ekohl the parameter ordering is still wrong

**Comment by ekohl** [comment]
> @ekohl the parameter ordering is still wrong

Yes, that's why I kept the ```= undef``` in the initial version. I don't have much time to complete this.

**Comment by bastelfreak** [comment]
hey @ekohl, are you still interested in this?

**Comment by ekohl** [comment]
I think this is still a good idea but I don't have time to finish this. I've also come around on the common.yaml and I now prefer defaults where possible in the .pp file because of puppet-strings.

**Comment by ekohl** [comment]
I don't have time to finish this.

**Pull request closed without merging**